### ISSUE TITLE
A deep copy occurs in the Drawable member draw function parameter RenderStates

### ISSUE THREAD

**Issue by Nrtyck** [SFML/SFML#3073]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I am here to report an issue and not to just ask a question or look for help (use the [forum](https://en.sfml-dev.org/forums/index.php#c3) or [Discord](https://discord.gg/nr4X7Fh) instead)

### Describe your issue here

![image](https://github.com/SFML/SFML/assets/37080373/aade6917-bad8-4e31-8e56-6cb832e50d16)
![image](https://github.com/SFML/SFML/assets/37080373/74b98e64-e860-4ae4-9b3c-f06c2181cc9c)

Whether to use `const RenderStates& states`, copy on demand when rendering?
 

### Your Environment

- OS / distro / window manager:  windows 11
- SFML version:  master 
- Compiler / toolchain: VS 2022
- Special compiler / CMake flags:

### Steps to reproduce

1. 
2. 

```cpp

```

### Expected behavior

![image](https://github.com/SFML/SFML/assets/37080373/66ef1e38-fee7-45ec-bb33-6706c96336da)
![image](https://github.com/SFML/SFML/assets/37080373/e030ea6b-ab8c-4cc1-8a42-f917664c33d5)

### Actual behavior

![image](https://github.com/SFML/SFML/assets/37080373/74b98e64-e860-4ae4-9b3c-f06c2181cc9c)

**Comment by Bambo-Borris** [comment]
I'd want to see some performance data here before we went down this route again. The drawable API survived many years with this copy instead of a reference, and because of it a common pattern of use occurred. The temporary move away to a reference meant a long standing way of making custom drawables didn't work. In fact, anyone making a custom drawable that they wanted to texture or apply a shader to would have to deep copy anyway, so the reference argument was barely saving anything in that case.

**Comment by vittorioromeo** [comment]
I still think it's worthwhile to look into this, the fact that we're needlessly copying the entirety of  `sf::RenderStates` even when not needed suggests that maybe there's an improvement to our API that we can make, even if the performance gain is not significant.

My intuition is that we can create a `sf::RenderStates` object at the beginning of our draw call chain, and then pass it as a non-`const` reference to all subsequent calls, thus getting mutation and preventing deep copies at the same time.

**Comment by Bambo-Borris** [comment]
> I still think it's worthwhile to look into this, the fact that we're needlessly copying the entirety of `sf::RenderStates` even when not needed suggests that maybe there's an improvement to our API that we can make, even if the performance gain is not significant.
> 
> My intuition is that we can create a `sf::RenderStates` object at the beginning of our draw call chain, and then pass it as a non-`const` reference to all subsequent calls, thus getting mutation and preventing deep copies at the same time.

This adds implications down the chain. If an object updates the shader, and the next one does not remove it, then the next in the chain will be drawn with the shader of the previous. The same applies for the texture instance or the coordinate type you apply. Also, under the const reference approach any custom drawable that alters the render states has to do the deep copy anyway.

**Comment by vittorioromeo** [comment]
> > I still think it's worthwhile to look into this, the fact that we're needlessly copying the entirety of `sf::RenderStates` even when not needed suggests that maybe there's an improvement to our API that we can make, even if the performance gain is not significant.
> > My intuition is that we can create a `sf::RenderStates` object at the beginning of our draw call chain, and then pass it as a non-`const` reference to all subsequent calls, thus getting mutation and preventing deep copies at the same time.
> 
> This adds implications down the chain. If an object updates the shader, and the next one does not remove it, then the next in the chain will be drawn with the shader of the previous. The same applies for the texture instance or the coordinate type you apply. Also, under the const reference approach any custom drawable that alters the render states has to do the deep copy anyway. And by the time this copy is a performance concern, you're probably drawing so many things you have a much bigger problem with the immediate mode rendering approach SFML currently uses, and that will be eating the biggest amount of your performance.

Isn't that the desired behavior? If a drawable's draw call ends up drawing some drawable children, they should inherit the render states of the parent.

**Comment by Bambo-Borris** [comment]
> > > I still think it's worthwhile to look into this, the fact that we're needlessly copying the entirety of `sf::RenderStates` even when not needed suggests that maybe there's an improvement to our API that we can make, even if the performance gain is not significant.
> > > My intuition is that we can create a `sf::RenderStates` object at the beginning of our draw call chain, and then pass it as a non-`const` reference to all subsequent calls, thus getting mutation and preventing deep copies at the same time.
> > 
> > 
> > This adds implications down the chain. If an object updates the shader, and the next one does not remove it, then the next in the chain will be drawn with the shader of the previous. The same applies for the texture instance or the coordinate type you apply. Also, under the const reference approach any custom drawable that alters the render states has to do the deep copy anyway. And by the time this copy is a performance concern, you're probably drawing so many things you have a much bigger problem with the immediate mode rendering approach SFML currently uses, and that will be eating the biggest amount of your performance.
> 
> Isn't that the desired behavior? If a drawable's draw call ends up drawing some drawable children, they should inherit the render states of the parent.

Maybe in an object which is composed of multiple sub-drawables, but what of in the case of a chain of calls to a mix of shapes and sprites. If you drew the first one with a shader, but didn't want the rest you're forced to know to nullify the shader argument. You effectively must manually reset the render states back for the parameters concerned... Or the more likely way of doing it is a user will just reconstruct the render states between draws anyway. Bringing us back to the deep copy here. Remember if a user supplies no `sf::RenderStates`, then there's a default one supplied down the chain. Making it so the early things drawn have implications for the later ones sounds like a way to cause problems on downstream draws. Users will be drawing front-to-back also, so the background would have implications on the foreground items if it alters the render states and they aren't reset back.

**Comment by eXpl0it3r** [comment]
There's no point in trying to come up with a completely new rendering API at this point - it's not the priority of SFML 3.

SFML 4 will feature a complete change in rendering API, so investing time into SFML 3, which focusses on bringing C++17 support, isn't worth it and a best just alienates SFML 2 users from using SFML 3.

I only kept this open as to give @Nrtyck time to respond in case I've misunderstood them. We've decided not to go forth with the change and the decision doesn't need rediscussing every few months. 😉
